### PR TITLE
bpf: add -nostdinc and a few more misc compilation options

### PR DESCRIPTION
**Pull request by borkmann**
See commit msg.

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

File: bpf/lib/policy_log.h [hunk 1]
```diff
@@ -47,30 +47,31 @@ send_policy_verdict_notify(struct __ctx_buff *ctx, __u32 remote_label, __u16 dst
...
+	msg.match_type = match_type;
+	msg.verdict = verdict;
+
+	msg.dst_port = bpf_ntohs(dst_port);
+	msg.proto = proto;
+	msg.dir = dir;
+	msg.ipv6 = is_ipv6;
+
+	msg.pad0 = 0;
+	msg.pad1 = 0;
```
Comment (joestringer) [inline bpf/lib/policy_log.h]: The nice thing about the previous initializer was that we get auto-zeroed fields if it's extended and a developer forgets to set the field, whereas now it's easier to break right?

File: bpf/lib/trace.h [hunk 2]
```diff
@@ -151,29 +155,28 @@ send_trace_notify(struct __ctx_buff *ctx, __u8 obs_point, __u32 src, __u32 dst,
...
-		.ifindex = ifindex,
-		.orig_ip4 = 0,
-		.orig_pad1 = 0,
-		.orig_pad2 = 0,
-		.orig_pad3 = 0,
-	};
+	msg.type = CILIUM_NOTIFY_TRACE;
+	msg.subtype = obs_point;
+	msg.source = EVENT_SOURCE;
+	msg.version = NOTIFY_CAPTURE_VER;
```
Comment (joestringer) [inline bpf/lib/trace.h]: can't help but wonder if there's some better code refactoring/sharing we could do when reviewing ~three~ four separate, very similar, initializers like this ;)

**Review by joestringer — APPROVED** [review]
Couple minor nits but changes LGTM.

File: bpf/Makefile.bpf [hunk 3]
```diff
@@ -3,11 +3,14 @@
 
 FLAGS := -I$(ROOT_DIR)/bpf/include -I$(ROOT_DIR)/bpf -D__NR_CPUS__=$(shell nproc) -O2 -g
 
-CLANG_FLAGS := ${FLAGS} -target bpf -emit-llvm
+CLANG_FLAGS := ${FLAGS} -target bpf -std=gnu89 -nostdinc -emit-llvm
```
Comment (pchaigno) [inline bpf/Makefile.bpf:6]: Do we need the same flags in `test/bpf/Makefile`?

File: bpf/lib/trace.h (hunk 2 above)
Comment (borkmann) [reply bpf/lib/trace.h]: I'll make it a helper to simplify all the other types like this and address the other comment from above.

File: bpf/Makefile.bpf (hunk 3 above)
Comment (borkmann) [reply bpf/Makefile.bpf:6]: Yeah probably makes sense, will add, the CFLAGS lack also few other things in there.

**Comment by borkmann** [comment]
test-me-please

File: bpf/lib/dbg.h [hunk 4]
```diff
@@ -148,34 +149,28 @@ struct debug_msg {
 static __always_inline void cilium_dbg(struct __ctx_buff *ctx, __u8 type,
 				       __u32 arg1, __u32 arg2)
 {
-	__u32 hash = get_hash_recalc(ctx);
```
Comment (qmonnet) [inline bpf/lib/dbg.h:151]: All the `get_hash_recalc()` removal should be moved to first commit?

**Review by qmonnet — COMMENTED** [review]
Looks like removal of `get_hash_recalc()` got split between the two patches.

Otherwise looks good, but second patch is huge. Did all of those fixes came out of the new warnings at compile time?

**Comment by borkmann** [comment]
test-me-please

File: bpf/lib/dbg.h (hunk 4 above)
Comment (borkmann) [reply bpf/lib/dbg.h:151]: I moved it into the __notify_common_hdr() in the 2nd one, I'm also fine to remove the macro entirely, no strong opinion.

**Comment by borkmann** [comment]
> Otherwise looks good, but second patch is huge. Did all of those fixes came out of the new warnings at compile time?

Yeah not all of it, for example, the event output msg refactoring touches all other ones as well to have a consistent initialiser. Hmm, I can split if off from the 2nd one to shrink it further.

**Comment by qmonnet** [comment]
> Hmm, I can split if off from the 2nd one to shrink it further.

If you don't mind too much, that would help with the review.

**Comment by borkmann** [comment]
test-me-please

File: bpf/lib/dbg.h (hunk 4 above)
Comment (qmonnet) [reply bpf/lib/dbg.h:151]: Ah right it's moved, not removed. Ok it's fine as is then, sorry.

File: bpf/lib/trace.h [hunk 5]
```diff
@@ -143,112 +143,87 @@ static __always_inline void
 send_trace_notify(struct __ctx_buff *ctx, __u8 obs_point, __u32 src, __u32 dst,
 		   __u16 dst_id, __u32 ifindex, __u8 reason, __u32 monitor)
 {
+	__u64 ctx_len = ctx_full_len(ctx);
+	__u64 cap_len = min_t(__u64, !monitor ? TRACE_PAYLOAD_LEN : monitor,
```
Comment (qmonnet) [inline bpf/lib/trace.h]: Nit: `!monitor ? TRACE_PAYLOAD_LEN : monitor` is the same as `monitor ? : TRACE_PAYLOAD_LEN` I think, which might be more readable?

File: bpf/bpf_overlay.c [hunk 6]
```diff
@@ -113,7 +115,6 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
 	if (1) {
```
Comment (qmonnet) [inline bpf/bpf_overlay.c:115]: Not related to that PR, but I've been curious: Do you know what those `if (1)` are for?

**Review by qmonnet — APPROVED** [review]
Looks great!

File: bpf/lib/trace.h (hunk 5 above)
Comment (borkmann) [reply bpf/lib/trace.h]: Ah yeah of course, will fix :-)

**Comment by borkmann** [comment]
test-me-please

File: bpf/bpf_overlay.c (hunk 6 above)
Comment (borkmann) [reply bpf/bpf_overlay.c:115]: My assumption is to get a new scope for declaring vars, here host_mac and router_mac. Could also have been plain `{ ... }`to make it more clear.

**Comment by borkmann** [comment]
```
13:53:00  echo "docker push cilium/docker-plugin:latest-amd64"
13:53:00  docker push cilium/docker-plugin:latest-amd64
13:53:00  build -f hubble-relay.Dockerfile -t "cilium/hubble-relay:latest" .
13:53:00  /bin/bash: build: command not found
13:53:00  Makefile:290: recipe for target 'docker-hubble-relay-image' failed
13:53:00  make: *** [docker-hubble-relay-image] Error 127
Post stage
```

**Comment by borkmann** [comment]
test-me-please

**Comment by qmonnet** [comment]
https://cilium.slack.com/archives/C7PE7V806/p1588256186289000

**Comment by qmonnet** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
restart-ginkgo

**Comment by borkmann** [comment]
test-focus K8sService.*

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-focus K8sService.*

**Comment by borkmann** [comment]
```
22:53:35      runtime: already cached: 	docker.io/library/memcached:1.5.11
22:53:35      runtime: already cached: 	docker.io/tgraf/netperf:v1.0
22:53:35  Cancelling nested steps due to timeout
22:53:35  Sending interrupt signal to process
22:53:35      runtime: already cached: 	docker.io/wurstmeister/kafka:2.11-0.11.0.3
22:53:35      runtime: not cached in VM: 	quay.io/cilium/cilium-builder:2020-04-16
22:53:35      runtime: pulling: 		quay.io/cilium/cilium-builder:2020-04-16
22:53:36  script returned exit code 255
```

**Comment by borkmann** [comment]
restart-ginkgo

**Comment by qmonnet** [comment]
restart-ginkgo

**Comment by aanm** [comment]
@borkmann it seems Cilium was in CrashLoopback mode. Doesn't seem to be a flake

**Comment by borkmann** [comment]
> @borkmann it seems Cilium was in CrashLoopback mode. Doesn't seem to be a flake

Yep, saw it, will keep working on it in the evening today.

**Comment by borkmann** [comment]
test-focus K8sService.*

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
provision error

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
4.19:
```
17:10:05  + python get-gh-comment-info.py test-me-please --retrieve=version
17:10:05  + sed s/^$/419/
17:10:05  + ./print-node-ip.sh
17:10:05  + KERNEL=419 CILIUM_REGISTRY=147.75.199.49 ./vagrant-ci-start.sh
17:10:05  destroying vms in case this is a retry
17:10:05  The provider 'virtualbox' that was requested to back the machine
17:10:05  'k8s1-1.17' is reporting that it isn't usable on this system. The
17:10:05  reason is shown below:
17:10:05  
17:10:05  VirtualBox is complaining that the kernel module is not loaded. Please
17:10:05  run `VBoxManage --version` or open the VirtualBox GUI to see the error
17:10:05  message which should contain instructions on how to fix this error.
[Pipeline] }
```

**Comment by borkmann** [comment]
test-with-kernel

**Comment by borkmann** [comment]
test-focus K8sService.*

**Comment by borkmann** [comment]
test-me-please

**Pull request merged**